OpenApiQuery.EntityFramework/EntityFrameworkMaterializer.cs

```diff
@@ -0,0 +1,20 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+
+namespace OpenApiQuery.EntityFramework
+{
+    public struct EntityFrameworkMaterializer : IMaterializer
```

Comment:
The more usual pattern in the ASP.net core world is that `IMaterializer` is a service registered on DI. The extension libraries like this one, would provide an extension method to allow configuration like this: 

```cs
services.AddOpenApiQuery() // registers default services, returns an IOpenApiQueryBuilder
    .AddEntityFramework() // registers the entity framework materializer using this library
``` 

Within a single product/service you usually rather not need to specify the materializer as part of every query, you register the implementation once. If you still need flexibility, you still can register a custom service that uses other services like the IHttpContextAccessor to decide case-by-case.

Reply:
The point of `IMaterializer` is it's a typeclass. It doesn't and doesn't have to be registered anywhere because you don't want to create an instance of it. You only want to access static methods. This is why `struct` constraint exist. 

Again, you don't need to register `IMaterializer` typeclass. It's basically a marker which static method exist for the type. Unfortunately, C# doesn't provide better way to do it, but [it's the default encoding](https://github.com/dotnet/csharplang/issues/164) (see Implementation paragraph).

Reply:
I hear of this type class concept the first time sounds like a nice feature for some scenarios, but I think it does not apply here very well. IMO IMaterializer should be a DI service and should be registered as singleton. As a user of this library you want to configure the behavior of this library once at a central location. It sounds to me rather like an uncommon use case to vary the Materializer at every API endpoint.

Also the concept of these Type-Classes seem to be not very common in the ASP.net core world. I haven't seen this before. For the sake of simplicity for developers using this library I would personally rather go for the classical DI service approach.

Reply:
But you can configure it as you wish, when you write your controller. For example:

```cs
[HttpGet("order")]
public async Task<OpenApiQueryResult<Order>> GetAsync(
	OpenApiQueryOptions<Order, MongoMaterializer> queryOptions,
	CancellationToken cancellationToken)
{
	...
}
```

or


```cs
[HttpGet("order")]
public async Task<OpenApiQueryResult<Order>> GetAsync(
	OpenApiQueryOptions<Order, EntityFrameworkMaterializer> queryOptions,
	CancellationToken cancellationToken)
{
	...
}
```

Or whatever you want. You don't need DI to specify types of your input, C# compiler is excellent at this. As you can see there is an inversion of control, so we chose the implementation on call site while caller only know the `IMaterializer` interface.